keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+
+With the new methods we would like also to migrate to klog v2 and introduce a new JSON logging output format to Kubernetes components that would be an alternative to the current text format and which will make querying and processing Kubernetes logs even simpler.
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
+
+We would like for the Kubernetes community to settle on one preferred log message structure. Proposed structure should:
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using “%q” fmt logic
+* keys are formatted using “%s” fmt logic
+* values are formatted using “%v” fmt logic and optionally escaped using the “%q” fmt logic if it not a number or a boolean
+* keys are separated from values with a “=” sign
+* message and key-values are separated with a single space
+
+#### New klog methods
+
+To enforce a new log message structure we want to introduce new methods to klog library which will provide a more structured interface for formatting log messages compared to current methods based on fmt format strings.
+For each format method (`Infof`, `Warningf`, `Errorf`, `Fatalf`) we will add matching structured method (`InfoS`, `WarningS`, `ErrorS`, `FatalS`).
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] “Pod status updated“ pod=“kubedns“ status=“ready“
+```
+
+### References to Kubernetes objects
+
+Currently there is no consistency when logging references to Kubenetes objects like Pods. Depending on a log just the name of a Pod, the name and namespace as separate fields or Pod name prefixed with a namespace and ‘/’ character is logged.
+This makes finding logs related to a given object much harder.
+To address this problem we want to implement helper functions for all standard object types which could be used to build values which when logged will provide a consistent identification of each of those objects.
+For instance for namespaced objects those methods should produce following value: “<namespace>/<name>”.
+
+Example:
+```go
+klog.InfoS("Pod status updated", "pod", metadata.Pod(pod), "status", "ready")
```

Comment:
I'm not a fan of `metadatat.Pod()`.

Given that all objects implement runtime.Object interface, at the very least it shouldn't require passing `Pod()` part - we can just use the generic accessor.
But I would actually go even one step further and say that if runtime.Object is passed, we always appropriate print <namespace>/<name> by default, and if you really want to print the whole object, you need to do that on your own and pass the stringified-version of the object yourself (alternatively expose some helper for it).

Reply:
Good suggestion!
Usage of additional functions to extract metadata will be error prone. Good default could make big difference here.
I think it makes sense for `InfoS` to be able to extract metadata by itself. Klog v2 internally uses logr interface. Metadata extraction could be implemented as a step before passing arguments to logr. 

@thockin do you think it makes sense if we would be considering migration to logr in future? 
For that we could move metadata extraction logic to a wrapper for logr interface.

Reply:
I agree that this might be a good idea however it is not clear for me what should happen when an Kubernetes object is nested in some struct which is logged.
If in such case we would like also to log only a reference to an object and not the object itself then this becomes a serialization mechanism which will require more complex logic to handle and might add some overhead at runtime.
If this works only for top level objects then it will be probably easier to implement and more efficient but at the same time it might lead to mistakes.
One more thing to consider are debug logs (klog.v(10) logs) in which I could imagine that the full object might be logged. In such case we would need some mechanism to disable the implicit transformation.

Reply:
We shouldn't try to solve all the problems in the whole world. Yes - it doesn't fully solve problem of object nested in some other struct, but that's probably <1% of cases. So let's ignore them for now and solve those 99%. And we can get back to it in the future.